rfd/XXXX-tsh-upgrade.md

```diff
@@ -0,0 +1,117 @@
+---
+# RFD XXXX - tsh upgrade
+
+## Required Approvers
+
+* Engineering: ...
+* Infra: ...
+* Kube: ...
+
+## What
+
+A tsh subcommand that will prompt the user to upgrade if the version of tsh is older than the cluster version.
+
+## Why
+
+One of our goals for Teleport Cloud is to increase the cadence at which clusters can be upgraded.
+This requires that Teleport Agents and Teleport client tools be up to date with the latest compatible version of the cluster.
+
+Manually upgrading Teleport Agents and client tools can be a tedious task, so customers can end up delaying their upgrades.
+Simplifying this process will help customers upgrade and stay up to date with their Teleport version.
+
+Development for Automatic Upgrades of Teleport Agents has already been completed, so the success criteria for this RFD only includes developing a simpler way to keep Teleport client tools up to date.
+
+## Details
+
+The main idea here is to add an upgrade command to the Teleport client tools. This will detect the version of the Teleport cluster.
+If the version of tsh is older than the version of the cluster, the user will be prompted to upgrade.
+The command should then re-download and re-install itself.
+
+### Identifying cluster version
+
+Detecting the version of the Teleport cluster should be straight forward.
+Teleport supports a ping RPC that returns a response containing information about the Teleport version of the cluster and the minimum client version required.
+In the case that the cluster version of Teleport is newer than the tsh version, the user will be prompted to upgrade.
+
+### Identifying installation method
+
+The Teleport client tools should be re-installed using the original installation method.
+Teleport documents several different methods for installing the client tools. https://goteleport.com/docs/installation/
+- Teleport can be installed from source and built using go. This scenario can be handled the same as if Teleport was installed by downloading the pre-build binaries.
+- Teleport pre-built binaries can be downloaded for the desired platform.
+- Teleport can be installed through the system package managers. Teleport currently maintains DEB and RPM repositories.
+- Teleport is available for installation using brew on macOS, but it is not officially supported by Teleport and not recommended.
+- Teleport provides macOS installation packages. One package installs all Teleport binaries, while the other installs only the tsh client.
+- Teleport provides a tsh only zip installation for windows.
+
+Additionally, Teleport supports a couple different releases: Open Source, Enterprise, and a FIPS compliant release.
+The upgrade process will need to detect which release was installed and make sure to re-install the same release.
+
+In order to identify which installation method was used to install the Teleport client tools, various checks will be executed.
+The default installation method will be downloading the pre-built binaries.
```

Comment:
Having a different process per installation method sounds tedious. We'll have to maintain one installation/update method in tsh for each system/installation-method/package combination. Extensive testing will be hard and we'll have a lot of weird edge cases.

Is there a way to unify the update process and make it distribution-agnostic? e.g. storing `tsh` binaries in the user's home and installing a launcher system-wide.